### PR TITLE
III-4497 changes calendarType Single

### PR DESCRIPTION
**Pull request by JonasVHG**
### Changed

- Split up `SmallSingleFormatter` & `ExtraSmallSingleFormatter` cause they no longer return they same.
- Show year on `sm` when it is not the current year
- use relative dates for `today`, `tonight`, `tomorrow`, `this week`
- Use full year, full month and abbreaviated weekday on `md`

---

Ticket: https://jira.uitdatabank.be/browse/III-4497

### CODE REVIEW

File: src/DateComparison.php [hunk 1]
```diff
@@ -17,6 +18,33 @@ public static function onSameDay(DateTimeImmutable $date1, DateTimeImmutable  $d
         return $date1->format('Y-m-d') === $date2->format('Y-m-d');
     }
 
+    public static function isThisEvening(DateTimeImmutable $date): bool
+    {
+        return self::onSameDay($date, new DateTimeImmutable()) && (int)$date->format('G') >= 18;
```
Comment (bertramakers) [inline src/DateComparison.php]: I suspect these new methods should also make sure the timezone on the `DateTimeImmutable` is set to the default time zone. Otherwise a `DateTimeImmutable` in a timezone from a different part of the world could be given as an arugment and it would not be considered "this evening" because it's for example at 10am in that timezone, while it is in fact after 6pm in our timezone. (See `onSameDay()` in this same class)

You can also add unit tests for that by injecting a `DateTimeImmutable` on which you first call `setTimezone()` with a timezone of for example New York or Tokyo or something like that.

File: tests/Single/SmallSingleHTMLFormatterTest.php [hunk 2]
```diff
@@ -35,7 +36,101 @@ public function testFormatHTMLSingleDateXsOneDay(): void
...
+        $event = new Offer(
+            OfferType::event(),
+            new Status('Available', []),
+            new BookingAvailability('Available'),
+            new DateTimeImmutable($thisSunday->format('Y-m-d') . 'T18:30:00+01:00'),
+            new DateTimeImmutable($thisSunday->format('Y-m-d') . 'T22:30:00+01:00')
+        );
+
+        $this->assertEquals(
+            '<span class="cf-meta">Deze</span> <span class="cf-days">zondag</span>',
```
Comment (bertramakers) [inline tests/Single/SmallSingleHTMLFormatterTest.php:121]: Won't this fail when running this test on a Saturday or Sunday? 🤔 If you use the Carbon library for PHP instead of the native DateTimeImmutable directly you could mock the current date & time for easier testing these relative scenario's (also in other similar tests). See https://carbon.nesbot.com/docs/#api-testing

By mocking the current date & time you wouldn't need to try to calculate a real date in the near future from the real current date & time, and you won't run into such edge cases. Because you could mock the current date & time as a specific datetime that's for example a Thursday, and then use a test date for the next Sunday (relative to that mocked Thursday).

Carbon just extends the native DateTimeImmutable, so we don't need to refactor all code in this library to use it. We can just use it in places where it's useful for easier testing, and keep using DateTimeImmutable in other places for now.

File: src/DateComparison.php (hunk 1 above)
Comment (lucwollants) [reply src/DateComparison.php]: Maybe this can benefit from Carbon https://github.com/briannesbitt/carbon?
That has a method to check this https://github.com/briannesbitt/Carbon/blob/87a5e6f250b00f4422b2859baf60fb2c5d543e97/src/Carbon/Carbon.php#L108

Comment (bertramakers) [reply src/DateComparison.php]: I'm not sure if that does the timezone conversion first. Also see my other comment, Carbon has already been added

Comment (JonasVHG) [reply src/DateComparison.php]: But the comparisons can indeed be made simpler with carbon, but you'll still have to take care of the timezones.

Comment (bertramakers) [reply src/DateComparison.php]: Right. But `onSameDay()` is not really in scope of this PR I think? It's an existing method that is not touched or used for these changes. It was just an example of a method where we already do the timezone conversion, which we also need in the new `isThisEvening()` etc (for which Carbon does not seem to have methods)

Comment (bertramakers) [reply src/DateComparison.php]: Ah you already changed it. Okay 👍

File: composer.json [hunk 3]
```diff
@@ -14,7 +14,8 @@
       "ext-intl": "*",
       "ext-json": "*",
       "delfimov/translate": "^2.4",
-      "psr/container": "^1.0"
+      "psr/container": "^1.0",
+        "nesbot/carbon": "^2.63"
```
Comment (bertramakers) [inline composer.json]: Small indentation issue here now

**Review by bertramakers — APPROVED** [review]

File: src/DateComparison.php [hunk 4]
```diff
@@ -4,17 +4,49 @@
...
 
+use Carbon\CarbonImmutable;
+use DateInterval;
 use DateTimeImmutable;
 use DateTimeZone;
 
 final class DateComparison
 {
-    public static function onSameDay(DateTimeImmutable $date1, DateTimeImmutable  $date2): bool
+    public static function onSameDay(DateTimeImmutable $date1, DateTimeImmutable $date2): bool
```
Comment (lucwollants) [inline src/DateComparison.php:14]: I think this class needs extensive unit tests now. It is very important this works as expected because all formatters rely on it.

Comment (JonasVHG) [reply src/DateComparison.php:14]: I tend to agree, but I think this is out of scope for CalendarSummary changes, but Unit Testing the entire class would be a good idea. I'll make a JIRA ticket for it ( https://jira.uitdatabank.be/browse/III-5102 ).

Comment (lucwollants) [reply src/DateComparison.php:14]: Although most methods are new, so I would say it's in the scope 😉. But a ticket is fine.

**Review by lucwollants — APPROVED** [review]

File: src/DateComparison.php (hunk 4 above)
Comment (bertramakers) [reply src/DateComparison.php:14]: Hmm yes I overlooked that a lot of the new code uses this `onSameDay()` method underneath. But we can do it in a follow-up ticket/PR since the new code won't actually be used in SAPI3/UDB3 until we make a new release of this library and update the version in those apps, so there's not a lot of risk.
